google/resource_endpoints_service.go

```diff
@@ -0,0 +1,297 @@
+										Type:     schema.TypeString,
+										Required: true,
+									},
+									"request_type": &schema.Schema{
+										Type:     schema.TypeString,
+										Required: true,
+									},
+									"response_type": &schema.Schema{
+										Type:     schema.TypeString,
+										Required: true,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			"dns_address": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"endpoints": &schema.Schema{
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"address": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func getOpenApiConfigSource(config_text string) servicemanagement.ConfigSource {
+	// We need to provide a ConfigSource object to the API whenever submitting a
+	// new config.  A ConfigSource contains a ConfigFile which contains the b64
+	// encoded contents of the file.  OpenAPI requires only one file.
+	configfile := servicemanagement.ConfigFile{
+		FileContents: base64.StdEncoding.EncodeToString([]byte(config_text)),
+		FileType:     "OPEN_API_YAML",
+		FilePath:     "heredoc.yaml",
+	}
+	return servicemanagement.ConfigSource{
+		Files: []*servicemanagement.ConfigFile{&configfile},
+	}
+}
+
+func getGrpcConfigSource(service_config, proto_config string) servicemanagement.ConfigSource {
+	// gRPC requires both the file specifying the service and the compiled protobuf,
+	// but they can be in any order.
+	yml_configfile := servicemanagement.ConfigFile{
+		FileContents: base64.StdEncoding.EncodeToString([]byte(service_config)),
+		FileType:     "SERVICE_CONFIG_YAML",
+		FilePath:     "heredoc.yaml",
+	}
+	proto_configfile := servicemanagement.ConfigFile{
+		FileContents: base64.StdEncoding.EncodeToString([]byte(proto_config)),
+		FileType:     "FILE_DESCRIPTOR_SET_PROTO",
+		FilePath:     "api_def.pb",
+	}
+	return servicemanagement.ConfigSource{
+		Files: []*servicemanagement.ConfigFile{&yml_configfile, &proto_configfile},
+	}
+}
+
+func resourceEndpointsServiceCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+	// If the service doesn't exist, we'll need to create it, but if it does, it
+	// will be reused.
+	serviceName := d.Get("service_name").(string)
+	servicesService := servicemanagement.NewServicesService(config.clientServiceMan)
+	_, err = servicesService.Get(serviceName).Do()
+	if err != nil {
+		_, err = servicesService.Create(&servicemanagement.ManagedService{ProducerProjectId: project, ServiceName: serviceName}).Do()
+		if err != nil {
+			return err
+		}
+	}
+	// Do a rollout using the update mechanism.
+	err = resourceEndpointsServiceUpdate(d, meta)
```

Comment:
Should it fail instead if the resource already exists? Like in all our other resources, if the resource already exists, the user should call terraform import.

Reply:
The "resource" here is really the rollout, I think - the creation of the service object is incidental to the creation of the first rollout.